gradio/routes.py

```diff
@@ -145,12 +145,16 @@ def get_blocks(self) -> gradio.Blocks:
             raise ValueError("No Blocks has been configured for this app.")
         return self.blocks
 
-    @staticmethod
-    def build_proxy_request(url_path):
+    def build_proxy_request(self, url_path):
         url = httpx.URL(url_path)
-        is_hf_url = url.host.endswith(".hf.space")
+        assert self.blocks
+        # Don't proxy a URL unless it's a URL specifically loaded by the user using
+        # gr.load() to prevent SSRF or harvesting of HF tokens by malicious Spaces.
+        is_safe_url = any(url.host.endswith(root) for root in self.blocks.root_urls)
+        if not is_safe_url:
+            raise ValueError("This URL cannot be proxied.")
         headers = {}
-        if Context.hf_token is not None and is_hf_url:
+        if Context.hf_token is not None:
```

Comment:
I think `is_hf_url` is still relevant, because someone could `gr.load()` from a non-hf-spaces URL.

Reply:
That's actually why I made this change. The idea is that only URLs that a developer trusts (because they have explicitly `gr.load()`-ed them) will be sent requests through `/proxy`

Reply:
I.e.
* In a Gradio app, a developer may call `gr.load()` to load external Gradio apps from Spaces or elsewhere. The URLs of these apps will be whitelisted
* Users of that Gradio app will only be able to call `/proxy` using the whitelisted URLs

Reply:
But why would a non-`.hf.space` origin need a `hf_token` authorization header?

Reply:
You're right, it wouldn't. I suppose we could still leave that there

Reply:
You can't gr.load a non-space though?

Reply:
@freddyaboulton the docs for `gr.load`, emphasis mine.

> Method that constructs a Blocks from a Hugging Face repo. Can **accept model repos (if src is "models") or** Space repos (if src is "spaces"). The input and output components are automatically loaded from the repo.

Reply:
@freddyaboulton is correct in that `gr.load()` can only be used to load Gradio apps from Spaces. You can use it with HF _model_ repos as well but the behavior is quite different (it creates a Gradio app which calls the HF Inference API under the hood). That being said @freddyaboulton I don't think this check hurts and may be useful in the future if we extend `gr.load()` to be able to load Gradio apps from arbitrary URLs similar to the gradio client.